ofrak_core/ofrak/cli/command/unpack.py

```diff
@@ -0,0 +1,118 @@
+import os.path
+from typing import Dict
+
+from ofrak import OFRAKContext, Resource
+
+import argparse
+from pathlib import Path
+import time
+import sys
+
+from ofrak.cli.command.identify import Identify
+from ofrak.cli.ofrak_cli import OfrakCommandRunsScript
+from ofrak.core import FilesystemEntry
+
+
+class Unpack(OfrakCommandRunsScript):
+    def create_parser(self, parser: argparse._SubParsersAction):
+        subparser = parser.add_parser(
+            "unpack",
+            help="Unpack all identified structures that can be unpacked with OFRAK",
+            description="Import a file as an OFRAK resource, then identifies and unpacks it. The "
+            "resource's children are written to the output directory as individual "
+            "files. Children which have no data are not written as files. A file `__ofrak_info__` "
+            "is also written to the output directory, containing the known OFRAK tags and "
+            "attributes for each descendant.",
+        )
+        subparser.add_argument(
+            "-o",
+            "--output_directory",
+            help="Directory to write unpacked resource tree to. If no directory is given, a new one"
+            " will be created in the same directory as the file being unpacked.",
+        )
+        subparser.add_argument(
+            "--print-info",
+            "-p",
+            help="Print contents of __ofrak_info__ (which may be large!) to stdout as well as the "
+            "__ofrak_info__ file.",
+            action="store_true",
+        )
+        subparser.add_argument("filename", help="File to unpack")
+        self.add_ofrak_arguments(subparser)
+
+        return subparser
+
+    async def ofrak_func(self, ofrak_context: OFRAKContext, args: argparse.Namespace):
+        print(f"Unpacking file: {args.filename}\n")
+        root_resource = await ofrak_context.create_root_resource_from_file(args.filename)
+
+        await root_resource.unpack()
+
+        if args.output_directory:
```

Comment:
My expectation when running `python3 -m ofrak unpack` is that I can get the following information:

1. See the unpacked tree `resource.summarize_tree`
2. I have a place where I can dig in to see a) what identifiers ran/got tagged b) what unpacker was run.

So for example, `python3 -m ofrak unpack /ofrak_core/test_ofrak/components/assets/binwalk_assets/dirtraversal.tar` would print out the tree:
```
┌a9f8076c82ee4015949203f80eb01b00: [caption=(File, TarArchive), attributes=(Magic), global_offset=(0x0-0x43800), parent_offset=(0x0-0x0), data_hash=88206ede]
├───┬1e1e5a4f8c494c078b2869bb84c49a8d: [caption=(Folder: .), attributes=(FilesystemEntryAutoAttributes), global_offset=(0x0-0x0), parent_offset=(0x0-0x0), data_ascii=""]
│   └────2a1c58b97aa942b0a064505e45aa854f: [caption=(SymbolicLink: ./foo), attributes=(SymbolicLinkAutoAttributes, FilesystemEntryAutoAttributes), global_offset=(0x0-0x0), parent_offset=(0x0-0x0), data_ascii=""]
├───┬dd139033910246bbb52f5f392bb79458: [caption=(Folder: subdir), attributes=(FilesystemEntryAutoAttributes), global_offset=(0x0-0x0), parent_offset=(0x0-0x0), data_ascii=""]
│   ├────880b84ae69c3472dba29a4d17ab95f44: [caption=(SymbolicLink: subdir/foo2), attributes=(SymbolicLinkAutoAttributes, FilesystemEntryAutoAttributes), global_offset=(0x0-0x0), parent_offset=(0x0-0x0), data_ascii=""]
│   ├────5263e0ac48f44dadb50e3092bdaac359: [caption=(File: README2.txt), attributes=(FilesystemEntryAutoAttributes), global_offset=(0x0-0x17), parent_offset=(0x0-0x0), data_hex=5375626469722074726176657273616c20746573742e0a]
│   ├────17a81359e28f4c73983491099598c54c: [caption=(SymbolicLink: subdir/README_link), attributes=(SymbolicLinkAutoAttributes, FilesystemEntryAutoAttributes), global_offset=(0x0-0x0), parent_offset=(0x0-0x0), data_ascii=""]
│   ├────417e7c1087144b39962167bd1f2463e1: [caption=(SymbolicLink: subdir/bar2), attributes=(SymbolicLinkAutoAttributes, FilesystemEntryAutoAttributes), global_offset=(0x0-0x0), parent_offset=(0x0-0x0), data_ascii=""]
│   └────d0e00b7fd0934b13b1355e08ed87de53: [caption=(File: ls2), attributes=(FilesystemEntryAutoAttributes), global_offset=(0x0-0x20aa0), parent_offset=(0x0-0x0), data_hash=0d06f972]
├────a2104e60df7b408c96b4a4f5ec250766: [caption=(File: README.txt), attributes=(FilesystemEntryAutoAttributes), global_offset=(0x0-0x13), parent_offset=(0x0-0x0), data_hex=4469722074726176657273616c20746573740a]
├────2e773718941545d0abbd6d868119111c: [caption=(SymbolicLink: README2_link), attributes=(SymbolicLinkAutoAttributes, FilesystemEntryAutoAttributes), global_offset=(0x0-0x0), parent_offset=(0x0-0x0), data_ascii=""]
├────3c50eab9d2594b09876c2c537b73b9d2: [caption=(SymbolicLink: bar), attributes=(SymbolicLinkAutoAttributes, FilesystemEntryAutoAttributes), global_offset=(0x0-0x0), parent_offset=(0x0-0x0), data_ascii=""]
└────435ddc9bcd3c4f2f9ace183012acb238: [caption=(File: ls), attributes=(FilesystemEntryAutoAttributes), global_offset=(0x0-0x20aa0), parent_offset=(0x0-0x0), data_hash=0d06f972]
```

and somehow, somewhere contain information indicating that the `TarUnpacker` had run.

Contract the above with the current output:
```
% python3 -m ofrak unpack dirtraversal.tar
Unpacking file: dirtraversal.tar

Extracting data to dirtraversal.tar_extracted_20230112104445
It took 0.040 seconds to run the OFRAK script
```
Running with `-p` creates unreadable output:
```
% python3 -m ofrak unpack -p dirtraversal.tar
Unpacking file: dirtraversal.tar

Extracting data to dirtraversal.tar_extracted_20230112104520
dirtraversal.tar_extracted_20230112104520/dirtraversal.tar
= Tags =
  File
  FilesystemEntry
  FilesystemRoot
  TarArchive
  GenericBinary
= Attributes =
  FilesystemEntryAutoAttributes(name=dirtraversal.tar, stat=os.stat_result(st_mode=33188, st_ino=103034880, st_dev=16777220, st_nlink=1, st_uid=501, st_gid=20, st_size=276480, st_atime=1673536699, st_mtime=1673535468, st_ctime=1673535468), xattrs={})  Magic:
   Mime: application/x-tar
   Descriptor: POSIX tar archive (GNU)


dirtraversal.tar_extracted_20230112104520/dirtraversal.tar.ofrak_children/.
= Tags =
  Folder
  FilesystemEntry
= Attributes =
  FilesystemEntryAutoAttributes(name=., stat=None, xattrs=None)

dirtraversal.tar_extracted_20230112104520/dirtraversal.tar.ofrak_children/..ofrak_children/./foo
= Tags =
  SymbolicLink
  FilesystemEntry
  SpecialFileType
= Attributes =
  FilesystemEntryAutoAttributes(name=./foo, stat=os.stat_result(st_mode=41453, st_ino=103043930, st_dev=16777220, st_nlink=1, st_uid=501, st_gid=20, st_size=6, st_atime=1673538320, st_mtime=1630850792, st_ctime=1673538320), xattrs={})  SymbolicLinkAutoAttributes(source_path=../../)

dirtraversal.tar_extracted_20230112104520/dirtraversal.tar.ofrak_children/subdir
= Tags =
  Folder
  FilesystemEntry
= Attributes =
  FilesystemEntryAutoAttributes(name=subdir, stat=os.stat_result(st_mode=16877, st_ino=103043934, st_dev=16777220, st_nlink=7, st_uid=501, st_gid=20, st_size=224, st_atime=1673538320, st_mtime=1630855557, st_ctime=1673538320), xattrs={})

dirtraversal.tar_extracted_20230112104520/dirtraversal.tar.ofrak_children/subdir.ofrak_children/subdir/foo2
= Tags =
  SymbolicLink
  FilesystemEntry
  SpecialFileType
= Attributes =
  FilesystemEntryAutoAttributes(name=subdir/foo2, stat=os.stat_result(st_mode=41453, st_ino=103043935, st_dev=16777220, st_nlink=1, st_uid=501, st_gid=20, st_size=12, st_atime=1673538320, st_mtime=1630851835, st_ctime=1673538320), xattrs={})  SymbolicLinkAutoAttributes(source_path=../../../../)

dirtraversal.tar_extracted_20230112104520/dirtraversal.tar.ofrak_children/subdir.ofrak_children/README2.txt
= Tags =
  File
  FilesystemEntry
= Attributes =
  FilesystemEntryAutoAttributes(name=README2.txt, stat=os.stat_result(st_mode=33188, st_ino=103043936, st_dev=16777220, st_nlink=1, st_uid=501, st_gid=20, st_size=23, st_atime=1673538320, st_mtime=1630855359, st_ctime=1673538320), xattrs={})

dirtraversal.tar_extracted_20230112104520/dirtraversal.tar.ofrak_children/subdir.ofrak_children/subdir/README_link
= Tags =
  SymbolicLink
  FilesystemEntry
  SpecialFileType
= Attributes =
  FilesystemEntryAutoAttributes(name=subdir/README_link, stat=os.stat_result(st_mode=41453, st_ino=103043937, st_dev=16777220, st_nlink=1, st_uid=501, st_gid=20, st_size=13, st_atime=1673538320, st_mtime=1630855557, st_ctime=1673538320), xattrs={})  SymbolicLinkAutoAttributes(source_path=../README.txt)

dirtraversal.tar_extracted_20230112104520/dirtraversal.tar.ofrak_children/subdir.ofrak_children/subdir/bar2
= Tags =
  SymbolicLink
  FilesystemEntry
  SpecialFileType
= Attributes =
  FilesystemEntryAutoAttributes(name=subdir/bar2, stat=os.stat_result(st_mode=41453, st_ino=103043938, st_dev=16777220, st_nlink=1, st_uid=501, st_gid=20, st_size=11, st_atime=1673538320, st_mtime=1630851846, st_ctime=1673538320), xattrs={})  SymbolicLinkAutoAttributes(source_path=/etc/shadow)

dirtraversal.tar_extracted_20230112104520/dirtraversal.tar.ofrak_children/subdir.ofrak_children/ls2
= Tags =
  File
  FilesystemEntry
= Attributes =
  FilesystemEntryAutoAttributes(name=ls2, stat=os.stat_result(st_mode=33261, st_ino=103043939, st_dev=16777220, st_nlink=1, st_uid=501, st_gid=20, st_size=133792, st_atime=1673538320, st_mtime=1630851750, st_ctime=1673538320), xattrs={})

dirtraversal.tar_extracted_20230112104520/dirtraversal.tar.ofrak_children/README.txt
= Tags =
  File
  FilesystemEntry
= Attributes =
  FilesystemEntryAutoAttributes(name=README.txt, stat=os.stat_result(st_mode=33188, st_ino=103043933, st_dev=16777220, st_nlink=1, st_uid=501, st_gid=20, st_size=19, st_atime=1673538320, st_mtime=1630855328, st_ctime=1673538320), xattrs={})

dirtraversal.tar_extracted_20230112104520/dirtraversal.tar.ofrak_children/README2_link
= Tags =
  SymbolicLink
  FilesystemEntry
  SpecialFileType
= Attributes =
  FilesystemEntryAutoAttributes(name=README2_link, stat=os.stat_result(st_mode=41453, st_ino=103043932, st_dev=16777220, st_nlink=1, st_uid=501, st_gid=20, st_size=18, st_atime=1673538320, st_mtime=1630855542, st_ctime=1673538320), xattrs={})  SymbolicLinkAutoAttributes(source_path=subdir/README2.txt)

dirtraversal.tar_extracted_20230112104520/dirtraversal.tar.ofrak_children/bar
= Tags =
  SymbolicLink
  FilesystemEntry
  SpecialFileType
= Attributes =
  FilesystemEntryAutoAttributes(name=bar, stat=os.stat_result(st_mode=41453, st_ino=103043929, st_dev=16777220, st_nlink=1, st_uid=501, st_gid=20, st_size=11, st_atime=1673538320, st_mtime=1630851650, st_ctime=1673538320), xattrs={})  SymbolicLinkAutoAttributes(source_path=/etc/passwd)

dirtraversal.tar_extracted_20230112104520/dirtraversal.tar.ofrak_children/ls
= Tags =
  File
  FilesystemEntry
= Attributes =
  FilesystemEntryAutoAttributes(name=ls, stat=os.stat_result(st_mode=33261, st_ino=103043931, st_dev=16777220, st_nlink=1, st_uid=501, st_gid=20, st_size=133792, st_atime=1673538320, st_mtime=1630851595, st_ctime=1673538320), xattrs={})
It took 0.038 seconds to run the OFRAK script
```

Reply:
I changed it to a modified version of the `summarize_tree`:

```
Extracting data to /tmp/blah
┌hello.rar: [attributes=(FilesystemEntryAutoAttributes, Magic), size=78 bytes, path=/tmp/blah/hello.rar]
└────hello.txt: [attributes=(FilesystemEntryAutoAttributes), size=11 bytes, path=/tmp/blah/hello.rar.ofrak_children/hello.txt]
```

Specifically, I have it print the filename instead of resource ID, also print out the full path where the resource has been dumped on disk, and stripped out all the stuff about the resource's data (you can get it trivially using the path) and replaced it with just a `size` for context.

Reply:
For context, here's what would have been printed above using the default summarize_tree:

```
┌0da2051eb66a4ca89dbe28bce68cd12d: [caption=(RarArchive, File: hello.rar), attributes=(FilesystemEntryAutoAttributes, Magic), global_offset=(0x0-0x4e), parent_offset=(0x0-0x0), data_hex=526172211a0701003392b5e50a010506000501018080006a104b951f02028b00068b00a483020d299d6185114a0d8000010968656c6c6f2e74787468656c6c6f20776f726c641d77565103050400]
└────ba421a623c714094be35861bf9c61105: [caption=(File: hello.txt), attributes=(FilesystemEntryAutoAttributes), global_offset=(0x0-0xb), parent_offset=(0x0-0x0), data_ascii="hello world"]
```

I think most of that is unhelpful, which is why I stripped it down.

Reply:
The modified `summarize_tree` output looks much more potentially useful.  One thing I would say is this seems most geared towards unpacking filesystems or archives. It might be helpful to think about unpacking binaries as well. For example, when unpacking an ELF, it would be helpful if the files created for each program header/section have the section name as their name. Or it might also be useful to specify where to stop unpacking (if I want a function inside a `CodeRegion`). Something like an xpath-like DSL might be the answer here. But this seems out of scope for the initial version, just something to think about.

Reply:
I think the caption should be added back to the summarize tree. The point of the caption is that it is a short string representation of the key attribute(s) of the resource. Removing that just to list the attribute class names seems worse for a user. 

To Sam's point, the caption would contain information about the section names.

Reply:
@rbs-jacob the name of the file itself is the caption (unless the file is actually a `FilesystemEntry`, in which case the actual filename that makes up most of the caption is used, and this also drops the "File: " prefix from the caption), so the caption in the summary string is redundant with the filename which replaces the ID. And, the sections' labels do not actually include section names since those have not been analyzed. As an example:

```
┌simple_arm_gcc.o.elf: [attributes=(FilesystemEntryAutoAttributes, Magic), size=34827 bytes, path=/tmp/bleh/simple_arm_gcc.o.elf]
├────ElfBasicHeader: [attributes=(ElfBasicHeaderAutoAttributes), size=16 bytes, path=/tmp/bleh/simple_arm_gcc.o.elf.ofrak_children/ElfBasicHeader]
├────ElfHeader: [attributes=(ElfHeaderAutoAttributes), size=36 bytes, path=/tmp/bleh/simple_arm_gcc.o.elf.ofrak_children/ElfHeader]
├────ElfSectionHeader: [attributes=(ElfSectionStructureAutoAttributes, ElfSectionHeaderAutoAttributes), size=40 bytes, path=/tmp/bleh/simple_arm_gcc.o.elf.ofrak_children/ElfSectionHeader]
├────ElfSection: [attributes=(ElfSectionStructureAutoAttributes), no data, (no path)]
├────ElfSectionHeader_1: [attributes=(ElfSectionStructureAutoAttributes, ElfSectionHeaderAutoAttributes), size=40 bytes, path=/tmp/bleh/simple_arm_gcc.o.elf.ofrak_children/ElfSectionHeader_1]
├────'ElfSection, CodeRegion': [attributes=(ElfSectionStructureAutoAttributes), size=24 bytes, path=/tmp/bleh/simple_arm_gcc.o.elf.ofrak_children/ElfSection, CodeRegion]
├────ElfSectionHeader_2: [attributes=(ElfSectionStructureAutoAttributes, ElfSectionHeaderAutoAttributes), size=40 bytes, path=/tmp/bleh/simple_arm_gcc.o.elf.ofrak_children/ElfSectionHeader_2]
├────'ElfSection, CodeRegion_1': [attributes=(ElfSectionStructureAutoAttributes), size=120 bytes, path=/tmp/bleh/simple_arm_gcc.o.elf.ofrak_children/ElfSection, CodeRegion_1]
├────ElfSectionHeader_3: [attributes=(ElfSectionStructureAutoAttributes, ElfSectionHeaderAutoAttributes), size=40 bytes, path=/tmp/bleh/simple_arm_gcc.o.elf.ofrak_children/ElfSectionHeader_3]
├────'ElfSection, CodeRegion_2': [attributes=(ElfSectionStructureAutoAttributes), size=24 bytes, path=/tmp/bleh/simple_arm_gcc.o.elf.ofrak_children/ElfSection, CodeRegion_2]
etc. etc.
```

Reply:
You can see that the sections do not include the `NamedProgramSectionAutoAttributes` attributes as these have not been analyzed. (You can also tell I haven't merged with master since they should now be called `AttributesType[NamedProgramSection]`)

Reply:
@EdwardLarson I like these changes!

What do you think about removing the unpack `-p` flag, and changing it so that running `python3 -m ofrak unpack [FILE]` **always** prints the summarized tree? I think this should be the default behavior--I can't really think of a usecase where I want to run the unpack command and not see the output tree.

Reply:
I also find the tree for decompressed filesystems a bit strange:

```
Extracting data to dirtraversal.tar_extracted_20230118150836
┌dirtraversal.tar: [attributes=(AttributesType[FilesystemEntry], Magic), size=276480 bytes, path=dirtraversal.tar_extracted_20230118150836/dirtraversal.tar]
├───┬.: [attributes=(AttributesType[FilesystemEntry]), size=0 bytes, (no path)]
│   └────./foo: [attributes=(AttributesType[FilesystemEntry], AttributesType[SymbolicLink]), size=0 bytes, (no path)]
├───┬subdir: [attributes=(AttributesType[FilesystemEntry]), size=0 bytes, (no path)]
│   ├────subdir/foo2: [attributes=(AttributesType[FilesystemEntry], AttributesType[SymbolicLink]), size=0 bytes, (no path)]
│   ├────README2.txt: [attributes=(AttributesType[FilesystemEntry]), size=23 bytes, path=dirtraversal.tar_extracted_20230118150836/dirtraversal.tar.ofrak_children/subdir.ofrak_children/README2.txt]
│   ├────subdir/README_link: [attributes=(AttributesType[FilesystemEntry], AttributesType[SymbolicLink]), size=0 bytes, (no path)]
│   ├────subdir/bar2: [attributes=(AttributesType[FilesystemEntry], AttributesType[SymbolicLink]), size=0 bytes, (no path)]
│   └────ls2: [attributes=(AttributesType[FilesystemEntry]), size=133792 bytes, path=dirtraversal.tar_extracted_20230118150836/dirtraversal.tar.ofrak_children/subdir.ofrak_children/ls2]
├────README.txt: [attributes=(AttributesType[FilesystemEntry]), size=19 bytes, path=dirtraversal.tar_extracted_20230118150836/dirtraversal.tar.ofrak_children/README.txt]
├────README2_link: [attributes=(AttributesType[FilesystemEntry], AttributesType[SymbolicLink]), size=0 bytes, (no path)]
├────bar: [attributes=(AttributesType[FilesystemEntry], AttributesType[SymbolicLink]), size=0 bytes, (no path)]
└────ls: [attributes=(AttributesType[FilesystemEntry]), size=133792 bytes, path=dirtraversal.tar_extracted_20230118150836/dirtraversal.tar.ofrak_children/ls]
```

Specifically, the `path` attribute seems to be providing information about the dumped directory and not about the actual files themselves: `subdir/README2.txt` has become `path=dirtraversal.tar_extracted_20230118150836/dirtraversal.tar.ofrak_children/subdir.ofrak_children/README2.txt]`.

Do we need to repeat the "ofrak_children" everywhere in the extracted folder? Is there a way for the `path` to actually represent the path of the file/resource within the original artifact?

Reply:
> Specifically, the path attribute seems to be providing information about the dumped directory and not about the actual files themselves

Yes, that's correct and intentional. Whether the dumped resources are the contents of a tar archive or ELF sections, the path printed there is so you can go on your local filesystem and open the file with the resource's data. 

> Is there a way for the path to actually represent the path of the file/resource within the original artifact?

Is it a sufficient answer that the `path` here is not supposed to be possible path within the original artifact? All resources will have a path to their local dump file, not all of them will have a path in some packed format context. 

> Do we need to repeat the "ofrak_children" everywhere in the extracted folder?

We could skip it for Folder resources, but then you would have an inconsistent structure where sometimes you go deeper by going into a ".ofrak_children" directory and sometimes you go down a named directory. For files that are unpacked and aren't actually directories, dumping them as files (not directories, even though they have children) preserves their data, so the children go in another folder. The dump is intended to just be a dump of the OFRAK resource tree, complete with all resource's data.

But, this brings up a good question about use cases. The `unpack` command was written and I have been writing it assuming that the goal is displaying the unpacked resource tree via the filesystem. Some of these comments are assuming that the goal is to display an unpacked filesystem on the local filesystem. Both are valid, but I think also have some opposing needs - for example, in the first use case, the data of a tar archive within a squashfs should be dumped to the filesystem and treated as a normal file, whereas in the second use case, that tar archive should I suppose just be treated as a directory.

Reply:
Maybe we should be splitting the command, or have a "filesystem mode" flag? Not sure. But either way, I don't think we will have just one style of "unpack" CLI, so we need to define the scope for this PR.

Reply:
I see. I think what is confusing is I expect `path` to be an artifact. What if you renamed `path` to `extracted-file`, `extracted-path` or something like that? That really is a path to how to get to the extracted file easily right?